storage/disk/disk_test.go

```diff
@@ -0,0 +1,1029 @@
+// Copyright 2021 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package disk
+
+import (
+	"bytes"
+	"context"
+	"strings"
+	"testing"
+
+	"github.com/open-policy-agent/opa/storage"
+	"github.com/open-policy-agent/opa/util"
+	"github.com/open-policy-agent/opa/util/test"
+)
+
+type testRead struct {
+	path string
+	exp  string
+}
+
+type testWrite struct {
+	op    storage.PatchOp
+	path  string
+	value string
+}
+
+type testWriteError struct {
+	op    storage.PatchOp
+	path  string
+	value string
+}
+
+func TestPolicies(t *testing.T) {
+
+	test.WithTempFS(map[string]string{}, func(dir string) {
+		ctx := context.Background()
+		s, err := New(ctx, Options{Dir: dir, Partitions: nil})
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		err = storage.Txn(ctx, s, storage.WriteParams, func(txn storage.Transaction) error {
+			ids, err := s.ListPolicies(ctx, txn)
+			if err != nil {
+				t.Fatal(err)
+			} else if len(ids) > 0 {
+				t.Fatal("unexpected policies found")
+			}
+
+			_, err = s.GetPolicy(ctx, txn, "foo.rego")
+			if err == nil {
+				t.Fatal("expected error")
+			}
+
+			err = s.DeletePolicy(ctx, txn, "foo.rego")
+			if err == nil {
+				t.Fatal("expected error")
+			}
+
+			err = s.UpsertPolicy(ctx, txn, "foo.rego", []byte(`package foo`))
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			ids, err = s.ListPolicies(ctx, txn)
+			if err != nil {
+				t.Fatal(err)
+			} else if len(ids) != 1 || ids[0] != "foo.rego" {
+				t.Fatalf("missing or unexpected policies found: %v", ids)
+			}
+
+			bs, err := s.GetPolicy(ctx, txn, "foo.rego")
+			if err != nil || !bytes.Equal(bs, []byte("package foo")) {
+				t.Fatalf("unexpected error or bad result: err=%v, result=%v", err, string(bs))
+			}
+
+			err = s.DeletePolicy(ctx, txn, "foo.rego")
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			ids, err = s.ListPolicies(ctx, txn)
+			if err != nil {
+				t.Fatal(err)
+			} else if len(ids) > 0 {
+				t.Fatal("unexpected policies found")
+			}
+
+			return nil
+		})
+		if err != nil {
+			t.Fatal(err)
+		}
+	})
+}
+
+func TestDataPartitioningValidation(t *testing.T) {
+
+	closeFn := func(ctx context.Context, s *Store) {
+		t.Helper()
+		if err := s.Close(ctx); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	test.WithTempFS(map[string]string{}, func(dir string) {
+
+		ctx := context.Background()
+
+		_, err := New(ctx, Options{Dir: dir, Partitions: []storage.Path{
+			storage.MustParsePath("/foo/bar"),
+			storage.MustParsePath("/foo/bar/baz"),
+		}})
+		if err == nil || !strings.Contains(err.Error(), "partitions are overlapped: [/foo/bar /foo/bar/baz]") {
```

Comment:
One thing that bothers me about `errors.As` is that I can't say `errors.As(err, &storage.Error{})` because of Go's type system; I need to say `var x interface{} = storage.Error{}; errors.As(err, &x)` to please the type checker.

I've improved the test code to assert on the code and message. It's better now.